### PR TITLE
172894510 anonymous student report

### PR DESCRIPTION
**Pull request by eireland**
Adds ability to open an anonymous student's report from the Generate Report button at the end of an activity. There was some work aslo done in activity-player, and report-services repos

### CODE REVIEW

File: js/actions/index.ts [hunk 1]
```diff
@@ -51,15 +53,34 @@ export const API_CALL = "API_CALL";
...
-      type: API_FETCH_PORTAL_DATA_AND_AUTH_FIRESTORE,
-      successAction: receivePortalData,
-      errorAction: fetchError,
-    },
-  };
+  const runKeyValue = queryValue("runKey");
+  if (runKeyValue) {
+    const activity= queryValue("activity") || "";
+    const source = activity? ((activity.split('/activities'))[0]).replace("https://","") : "";
+    const resourceUrl = queryValue("activity") || "";
```
Comment (pjanik) [inline js/actions/index.ts:60]: Seems like a duplicate of `activity` variable.

File: js/actions/index.ts [hunk 2]
```diff
@@ -51,15 +53,34 @@ export const API_CALL = "API_CALL";
...
+    const answerSource = queryValue("answerSource") || "";
+    return (dispatch: Dispatch, getState: any) => {
+      dispatch( {
+        type: SET_ANONYMOUS_VIEW,
+        runKey: runKeyValue
+      });
+      getState().get("report");
```
Comment (pjanik) [inline js/actions/index.ts:67]: I think this line doesn't do anything.

File: js/containers/portal-dashboard/portal-dashboard-app.tsx [hunk 3]
```diff
@@ -61,6 +61,7 @@ interface IState {
   initialLoading: boolean;
   scrollLeft: number;
   showAllResponsesPopup: boolean;
+  authError: string;
```
Comment (pjanik) [inline js/containers/portal-dashboard/portal-dashboard-app.tsx:64]: It doesn't seem to be used anywhere.

File: js/containers/portal-dashboard/portal-dashboard-app.tsx [hunk 4]
```diff
@@ -69,11 +70,12 @@ class PortalDashboardApp extends React.PureComponent<IProps, IState> {
     this.state = {
       initialLoading: true,
       scrollLeft: 0,
-      showAllResponsesPopup: false
+      showAllResponsesPopup: false,
+      authError: ""
```
Comment (pjanik) [inline js/containers/portal-dashboard/portal-dashboard-app.tsx:74]: It doesn't seem to be used anywhere.

File: js/containers/portal-dashboard/portal-dashboard-app.tsx [hunk 5]
```diff
@@ -69,11 +70,12 @@ class PortalDashboardApp extends React.PureComponent<IProps, IState> {
...
     };
   }
 
-  componentDidMount() {
+  async componentDidMount() {
```
Comment (pjanik) [inline js/containers/portal-dashboard/portal-dashboard-app.tsx:78]: `async` doesn't seem to be necessary here, there's no async code in the body of this function (no `await`).

File: js/util/url-query.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ */
+export const queryValue = (prop: string): string | undefined => {
+  const query = queryString.parse(window.location.search);
+  const val = query[prop];
+  if (!val) return;
+  if (Array.isArray(val)) {
+    throw `May only have one query parameter for ${prop}. Found: ${val}`;
+  }
+  return val;
+};
```
Comment (pjanik) [inline js/util/url-query.ts:16]: Note that we have urlParsing implemented in this repo, in `util/misc.ts`. It'd be good to reuse it if possible and/or add there more functions if they're useful.

File: js/util/url-query.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+export const setQueryValue = (prop: string, value: any, reload = false) => {
+  const parsed = queryString.parse(location.search);
+  parsed[prop] = value;
+  const newQueryString = queryString.stringify(parsed);
+  if (reload) {
+    location.search = queryString.stringify(parsed);
+  } else {
+    window.history.replaceState(null, "", "?" + newQueryString);
+  }
+};
```
Comment (pjanik) [inline js/util/url-query.ts:42]: These two functions don't seem to be used anywhere.

**Review by pjanik — CHANGES_REQUESTED** [review]
Generally looks good. I've left a few inline comments, but nothing major.

**Comment by eireland** [comment]
stale

File: js/actions/index.ts (hunk 1 above)
Comment (eireland) [reply js/actions/index.ts:60]: Good catch!

**Pull request closed without merging**